### PR TITLE
[10/n][enums/Sui] Update Sui package resolver for enums

### PR DESCRIPTION
**Pull request by tzakian**
## Description 

Adds support for Move enums to the Sui package resolver crate along with tests for it. 

This PR is meant to be a logically-reviewable portion but is not land-able on its own. It must be merged in with the changes in the rest of this stack to be landed.

## Stack:
* #17245
* #17246 
* #17247 
* #17248 
* #17249 **<< You are here**
* #17250 
* #17251 

## Test plan 

Added tests in the PR to make sure we can properly resolve enum types. 

---

## Release notes

Check each box that your changes affect. If none of the boxes relate to your changes, release notes aren't required.

For each box you select, include information after the relevant heading that describes the impact of your changes that a user might notice and any actions they must take to implement updates. 

- [ ] Protocol: 
- [ ] Nodes (Validators and Full nodes): 
- [ ] Indexer: 
- [ ] JSON-RPC: 
- [ ] GraphQL: 
- [ ] CLI: 
- [ ] Rust SDK:

### CODE REVIEW

**Review by tnowacki — APPROVED** [review]
image

File: crates/sui-package-resolver/src/snapshots/sui_package_resolver__tests__cross_module_enum.snap [hunk 1]
```diff
@@ -0,0 +1,23 @@
+---
+source: crates/sui-package-resolver/src/lib.rs
+expression: "format!(\"{layout:#}\")"
```
[the remark this answers is not in the document]
Comment (amnn) [reply crates/sui-package-resolver/src/snapshots/sui_package_resolver__tests__cross_module_enum.snap:3]: It's just the expression that generated the result no huge significance.

File: crates/sui-package-resolver/src/lib.rs [hunk 2]
```diff
@@ -101,6 +101,10 @@ pub struct Module {
     /// bytecode, to speed up definition lookups.
     struct_index: BTreeMap<String, (AccountAddress, StructDefinitionIndex)>,
 
+    /// Index mapping enum named to their defining ID and the index of their definition in the
```
Comment (amnn) [inline crates/sui-package-resolver/src/lib.rs:104]:
```suggestion
    /// Index mapping enum names to their defining ID and the index of their definition in the
```

File: crates/sui-package-resolver/src/lib.rs [hunk 3]
```diff
@@ -116,13 +120,40 @@ pub struct StructDef {
...
+    /// packages at their runtime IDs (not their storage ID or defining ID).
+    pub variants: Vec<VariantDef>,
+}
+
+/// Deserialized representation of an enum definition. These are always held inside an `EnumDef`.
+#[derive(Debug)]
+pub struct VariantDef {
+    /// The name of the enum variant
+    pub name: String,
+    /// The serialized representation of the variant's signature. Signatures refer to packages at
```
Comment (amnn) [inline crates/sui-package-resolver/src/lib.rs:152]: tiny nit, newline between fields with doc comments to match the rest in the file.

File: crates/sui-package-resolver/src/lib.rs [hunk 4]
```diff
@@ -197,6 +228,9 @@ pub enum OpenSignatureBody {
 struct ResolutionContext<'l> {
     /// Definitions (field information) for structs referred to by types added to this context.
     structs: BTreeMap<DatatypeKey, StructDef>,
+    /// Definitions (variant and field information) for enums referred to by types added to this
```
Comment (amnn) [inline crates/sui-package-resolver/src/lib.rs:231]: `ResolutionContext` thinks its special and it doesn't get newlines, but it's wrong.

File: crates/sui-package-resolver/src/lib.rs [hunk 5]
```diff
@@ -510,7 +544,7 @@ impl Package {
         let mut type_origins: BTreeMap<String, BTreeMap<String, AccountAddress>> = BTreeMap::new();
         for TypeOrigin {
             module_name,
-            struct_name,
+            datatype_name: struct_name,
```
Comment (amnn) [inline crates/sui-package-resolver/src/lib.rs:547]: nit: rename `struct_name` to `datatype_name`.

File: crates/sui-package-resolver/src/lib.rs [hunk 6]
```diff
@@ -565,10 +599,17 @@ impl Package {
         &self.modules
     }
 
+    /// Returns `true` if the package contains a definition for the struct with name `struct_name`.
+    fn is_struct_def(&self, module_name: &str, struct_name: &str) -> bool {
+        self.module(module_name)
+            .map(|module| module.struct_def(struct_name).is_ok_and(|x| x.is_some()))
+            .unwrap_or(false)
```
Comment (amnn) [inline crates/sui-package-resolver/src/lib.rs:606]: nit (optional)

```suggestion
            .is_some_and(|module| matches!(module.struct_def(struct_name), Ok(Some(_))))
```

File: crates/sui-package-resolver/src/lib.rs [hunk 7]
```diff
@@ -921,52 +1066,96 @@ impl<'l> ResolutionContext<'l> {
...
-                        TooManyTypeNodes, self.limits;
-                        max_type_nodes > self.structs.len()
-                    );
+                        check_max_limit!(
+                            TooManyTypeNodes, self.limits;
+                            max_type_nodes > self.structs.len()
+                        );
+
+                        self.structs.insert(key, def);
+                    } else {
```
[the remark this answers is not in the document]
Comment (amnn) [reply crates/sui-package-resolver/src/lib.rs:1116]: `T` is a `TypeTag` -- there is no `T::Enum`.

File: crates/sui-package-resolver/src/lib.rs [hunk 8]
```diff
@@ -1009,35 +1198,48 @@ impl<'l> ResolutionContext<'l> {
...
-                    let def = match self.structs.entry(key.clone()) {
-                        Entry::Occupied(e) => e.into_mut(),
+                    let type_params =
+                        if self.structs.contains_key(&key) || self.enums.contains_key(&key) {
+                            self.structs
+                                .get(&key)
+                                .map(|def| &def.type_params)
+                                .or(self.enums.get(&key).map(|def| &def.type_params))
+                                .expect("either struct or enum should be present")
+                        } else {
```
Comment (amnn) [inline crates/sui-package-resolver/src/lib.rs:1211]:
```suggestion
                        if let Some(def) = self.structs.get(&key) {
                            def.type_params
                        } else if let Some(def) = self.enums.get(&key) {
                            def.type_params
                        } else {
```

File: crates/sui-package-resolver/src/lib.rs
Comment (amnn) [inline crates/sui-package-resolver/src/lib.rs:1]: Reading over these changes really makes me wonder whether our lives would be simpler if structs were modeled as enums with only one variant (I mean just at the package resolver level, but maybe this applies to more places?) Seems like a lot of cases where we're checking two structures become one place.

Comment (amnn) [reply crates/sui-package-resolver/src/lib.rs:1]: By the time I reached the end of the file, I became fairly convinced about this. I think it's seriously worth considering (at least in terms of how `ResolutionContext` gathers type information -- the implementation would be a lot simpler if it had a single type for data type definitions, which it was gathering -- much less branching).

**Review by amnn — APPROVED** [review]
Accepting as this is all correct, but note the comment about trying to represent enums and structs in a unified way, at least in the `ResolutionContext`.

File: crates/sui-package-resolver/src/lib.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (tzakian) [reply crates/sui-package-resolver/src/lib.rs:1116]: Yea, it can either be a struct or an enum now, so it's not a great name. I really want to rename `TypeTag::Struct` to `TypeTag::Datatype`, but I'm going to hold off on that renaming till everything is in since that will just clutter a bunch of things up.

File: crates/sui-package-resolver/src/lib.rs
Comment (tzakian) [reply crates/sui-package-resolver/src/lib.rs:1]: Updated to only have one type for data definitions. However, within this type there is a field that branches based on struct/enum.

This is because I think keeping some level of branching helps when we go to construct type layouts, and in general this felt like the right tradeoff between sharing structure but also not overly sharing/making things "too loose".

File: crates/sui-package-resolver/src/lib.rs [hunk 9]
```diff
@@ -1384,32 +1573,17 @@ impl<'l> ResolutionContext<'l> {
...
-                let type_params = param_layouts.iter().map(|l| TypeTag::from(&l.0)).collect();
+                let type_params: Vec<TypeTag> =
+                    param_layouts.iter().map(|l| TypeTag::from(&l.0)).collect();
 
                 let type_ = StructTag {
                     address: def.defining_id,
                     module: ident(&key.module)?,
                     name: ident(&key.name)?,
-                    type_params,
+                    type_params: type_params.clone(),
```
Comment (amnn) [inline crates/sui-package-resolver/src/lib.rs:1583]: Couldn't figure out why these changes were necessary

File: crates/sui-package-resolver/src/lib.rs [hunk 10]
```diff
@@ -1288,29 +1428,78 @@ impl<'l> ResolutionContext<'l> {
...
                     type_params,
                 };
 
-                let mut fields = Vec::with_capacity(def.fields.len());
+                self.resolve_datatype_signature(def, type_, param_layouts, max_depth)?
+            }
+        })
+    }
+
+    fn resolve_datatype_signature(
```
Comment (amnn) [inline crates/sui-package-resolver/src/lib.rs:1447]: nit: could you add a doc comment here explaining what this function is doing?

File: crates/sui-package-resolver/src/lib.rs
Comment (amnn) [reply crates/sui-package-resolver/src/lib.rs:1]: Change looks good to me, thank you!
